pkg/pillar/nireconciler/linux_config.go

```diff
@@ -781,22 +815,20 @@ func (r *LinuxNIReconciler) getIntendedDnsmasqCfg(niID uuid.UUID) (items []dg.It
 	}
 
 	// DHCP server configuration
-	// By default, dnsmasq advertises a router (and we can have a static router
-	// defined in the NetworkInstanceConfig).
-	// To support airgap networks we interpret gateway=0.0.0.0 to not advertise
-	// ourselves as a router.
-	var gateway net.IP
-	if ni.bridge.Uplink.IfName != "" {
-		if ni.config.Gateway != nil {
-			if !ni.config.Gateway.IsUnspecified() {
-				gateway = ni.config.Gateway
-			} // else suppress router advertisement
-		} else if bridgeIP != nil {
-			gateway = bridgeIP.IP
-		}
-	} // else : Do not advertise router for air-gapped NI.
+	// dnsmasq advertises a router (default route) for network instance, unless:
+	//  a) network instance is air-gapped (without uplink)
+	//  b) uplink is app-shared and without default route
+	var gatewayIP net.IP
+	if bridgeIP != nil {
+		gatewayIP = bridgeIP.IP
+	}
+	var withDefaultRoute bool
+	airGap := ni.bridge.Uplink.IfName == ""
+	if !airGap && (ni.bridge.Uplink.IsMgmt || r.niHasDefRoute(ni)) {
```

Comment:
What's the need to check IsMgmt? Is this for a case when the management interface does not have a default route but somehow received all necessary prefix routes to reach the controller etc?

Note that the comment on L820 doesn't talk about IsMgmt.

Reply:
`IsMgmt  = ! app-shared`

This if-statement is a negation of that comment (we propagate default route if this if-statement is true).

Reply:
OK, but I still don't understand why you need to check IsMgmt

Reply:
For management uplink we always want to propagate default route to apps. For app-shared uplink (IsMgmt=false) only if the uplink has default route.

Reply:
Ah - your terminology might be confusing me.
We have two cases: inject a default route for the app instance (independent of routing on the external network)
propagate a default and/or other routes from the external network to the app instance.

Is it that the same code does the injection (for the management uplinks) and does the propagation?
If so it makes perfect sense, but is hard to understand from reading the code.

Reply:
This code here (`getIntendedDnsmasqCfg` function) only decides if app should receive default route for this network instance or not. If yes, then we propagate `0.0.0.0/0 via bridgeIP` - i.e. in this function we do not care what the actual gateway is and whether it is received from an external network or injected by user.
In function `getIntendedNIL3Cfg` we then prepare config for the NI routing table (for the host network stack). There we decide if we should copy default route from uplink routing table or put there a user-injected default route.